src/libraries/System.Text.Json/src/System/Text/Json/ThrowHelper.Serialization.cs

```diff
@@ -215,18 +216,37 @@ public static void ThrowInvalidOperationException_JsonPropertyRequiredAndExtensi
         }
 
         [DoesNotReturn]
-        public static void ThrowJsonException_JsonRequiredPropertyMissing(JsonTypeInfo parent, IEnumerable<int> missingJsonProperties)
+        public static void ThrowJsonException_JsonRequiredPropertyMissing(JsonTypeInfo parent, BitArray missingJsonProperties)
         {
             StringBuilder listOfMissingPropertiesBuilder = new();
             bool first = true;
 
             Debug.Assert(parent.PropertyCache != null);
 
-            foreach (int missingPropertyIndex in missingJsonProperties)
+            int propertyIdx = 0;
+
+            for (int requiredPropertyIdx = 0; requiredPropertyIdx < missingJsonProperties.Length; requiredPropertyIdx++)
```

Comment:
I know this is only meant to be invoked in exceptional circumstances, but it still is an $\mathcal O(n^2)$ algorithm that can be triggered by untrusted JSON inputs. I would strongly recommend making this linear-time (presumably by pre-computing an array of all required `JsonPropertyInfo` values).

Reply:
It's not `O(n^2)` - it's mostly (2*n) because we resume

Reply:
I'll simplify this code to make it clearer to see it's linear